### PR TITLE
Form enhancements

### PR DESCRIPTION
**Pull request by achilleas-k**
- Support Pre and Post Actions.  Pre actions can be used to set restrictions on forms such as values or enabling/disabling fields.  A PreAction can essentially redefine an entire form.
- Forms can now have multiple pages and support most input element types.
- Added documentation that describes the two built-in services.  This can serve as a guide for how to write more services.

### CODE REVIEW

**Comment by achilleas-k** [comment]
Added some tests for the form element rendering, slavishly trying to increase test coverage.

File: docs/labproject.md [hunk 1]
```diff
@@ -0,0 +1,39 @@
+# Lab project service
+
+The [lab project](/utonics/labproject/main.go) service takes advantage of almost all the features of Tonic.  Its purpose is to create repositories inside an organisation on GIN where the user is not an administrator or owner.
+It defines the following service components:
+
+## Form
+
+The Form consists of a two pages.
```
Comment (mpsonntag) [inline docs/labproject.md:8]:
```suggestion
The Form consists of two pages.
```

**Review by mpsonntag — APPROVED** [review]
:+1: for the dev doc.

File: docs/labproject.md [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+
+The first page consists of three elements:
+1. Lab organisation: The name of the organisation in which the repository is going to be created.
+2. Project name: The name of the repository or repositories to be created.
+3. Description: A long description for the project.
```
Comment (jcolomb) [inline docs/labproject.md:13]:
```suggestion
1. Lab organisation: The name of the organisation in which the repository is going to be created.
2. Optionally a team name where the repository will be added. If none is indicated, a team with the same name as the repository will be created.
3. Project name: The name of the repository or repositories to be created.Project name: The name of the repository or repositories to be created.
4. Description: A long description for the project.
```

Comment (jcolomb) [reply docs/labproject.md:13]: or is the documentation related to the existing service and this should be an enhancement, i.e. given in an issue ?

**Review by jcolomb — CHANGES_REQUESTED** [review]

File: docs/labproject.md (hunk 2 above)
Comment (achilleas-k) [reply docs/labproject.md:13]: No, that's fine.  I can add the field to the current PR, it's a small change.

**Comment by achilleas-k** [comment]
On second thought, @jcolomb please confirm the changes, lets merge with the doc change and I'll make the change to the code and form later.  There are a couple of things we might have to clarify regarding users and teams.

**Review by jcolomb — APPROVED** [review]

**Pull request merged**